spinup_bis/algos/tf2/bsp/sop.py

```diff
@@ -0,0 +1,512 @@
+        else:
+            return mu[ac_idx, 0]
+
+    @tf.function
+    def learn_on_batch(obs1, obs2, acts, rews, done):
+        with tf.GradientTape(persistent=True) as g:
+            # Actor update.
+            mu, _ = actor(obs1)
+            q_pi = critic1([obs1, mu])
+            pi_loss = -tf.reduce_mean(q_pi)
+
+            # Critic update.
+            q1 = critic1([obs1, acts])
+            q2 = critic2([obs1, acts])
+
+            _, pi_next = actor(obs2)
+            min_target_q = tf.minimum(
+                target_critic1([obs2, pi_next]),
+                target_critic2([obs2, pi_next]),
+            )
+
+            # Bellman backup for Q function.
+            q_backup = tf.stop_gradient(
+                rews + gamma * (1 - done) * min_target_q)
+            q1_loss = tf.reduce_mean((q_backup - q1) ** 2, axis=0)
+            q2_loss = tf.reduce_mean((q_backup - q2) ** 2, axis=0)
+
+            if use_weighted_bellman_backup:
+                q_std = tf.math.reduce_std(min_target_q, axis=0)
+                q_weight = tf.math.sigmoid(-q_std * bellman_temp) + 0.5
+
+                q_weight *= 0.5
+            else:
+                q_weight = 0.5
+
+            value_loss = tf.reduce_mean(q_weight * (q1_loss + q2_loss))
+
+        actor_gradients = g.gradient(pi_loss, actor.trainable_variables)
+        optimizer.apply_gradients(
+            zip(actor_gradients, actor.trainable_variables)
+        )
+        critic_gradients = g.gradient(value_loss, critic_variables)
+        optimizer.apply_gradients(
+            zip(critic_gradients, critic_variables))
+
+        for v, target_v in zip(critic1.trainable_variables,
+                               target_critic1.trainable_variables):
+            target_v.assign(polyak * target_v + (1 - polyak) * v)
+        for v, target_v in zip(critic2.trainable_variables,
+                               target_critic2.trainable_variables):
+            target_v.assign(polyak * target_v + (1 - polyak) * v)
+
+        del g
+        return dict(
+            pi_loss=pi_loss,
+            q1_loss=q1_loss,
+            q2_loss=q2_loss,
+            q1=q1,
+            q2=q2,
+        )
+
+    def test_agent():
+        for _ in range(num_test_episodes):
+            o, d, ep_ret, ep_len, task_ret = test_env.reset(), False, 0, 0, 0
+            while not (d or (ep_len == max_ep_len)):
+                o, r, d, info = test_env.step(
+                    evaluation_policy(tf.convert_to_tensor(o)))
+                ep_ret += r
+                ep_len += 1
+                task_ret += info.get('reward_task', 0)
+            logger.store(TestEpRet=ep_ret,
+                         TestEpLen=ep_len,
+                         TestTaskRet=task_ret,
+                         TestTaskSolved=info.get('is_solved', False))
+
+    def reset_episode(epoch):
+        o, ep_ret, ep_len, task_ret = env.reset(), 0, 0, 0
+        actor_idx = np.random.choice(ac_number)  # Select policy
```

Comment:
Remove this logic of sampling actor (you don't have ensemble anymore).

Reply:
done